connectionpool_systems_test.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2015 The gocql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+// +build conn_pool
+
+package gocql
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+// connection pool behavior test when nodes are removed from the cluster
+// to run this test, see connectionpool_systems_test.sh
+
+var (
```

Comment:
I'll add that the race condition in SimplePool seems to be between concurrent calls to ```Pick``` after the call to ```c.mu.Unlock()``` and another goroutine which is handling a node disconnect via the call to ```HandleError```. The intention is that when all connections for a host are removed, the host pool is removed, but since the later part of ```Pick``` isn't keeping the c.mu lock it seems to be able to get off a call to RoundRobin.Pick while another routine is holding the c.mu lock within removeConnLock which is modifying that same RoundRobin object, this can create a situation where Pick can return nil even if there are other valid connections to the cluster.

I didn't put any effort into fixing this race condition since the new policyConnPool obsoletes SimplePool, but if SimplePool is kept along side the new policyConnPool, then the race condition should be fixed.